File: src/main/java/com/alphawallet/attestation/SignedIdentityAttestation.java
package com.alphawallet.attestation;

import com.alphawallet.attestation.core.ASNEncodable;
import com.alphawallet.attestation.core.ExceptionUtil;
import com.alphawallet.attestation.core.SignatureUtility;
import com.alphawallet.attestation.core.Validateable;
import com.alphawallet.attestation.core.Verifiable;
import java.io.IOException;
import java.io.InvalidObjectException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1InputStream;
import org.bouncycastle.asn1.ASN1ObjectIdentifier;
import org.bouncycastle.asn1.ASN1Primitive;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DERBitString;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.ECPublicKeyParameters;

public class SignedIdentityAttestation implements ASNEncodable, Verifiable, Validateable {
  private static final Logger logger = LogManager.getLogger(SignedIdentityAttestation.class);

  public static final AlgorithmIdentifier ECDSA_WITH_SHA256 = new AlgorithmIdentifier(new ASN1ObjectIdentifier("1.2.840.10045.4.3.2"));

  private final IdentifierAttestation att;
  private final byte[] signature;
  private final AsymmetricKeyParameter attestationVerificationKey;

  public SignedIdentityAttestation(IdentifierAttestation att, AsymmetricCipherKeyPair attestationSigningKey) {
    this.att = att;
    this.signature = SignatureUtility.signWithEthereum(att.getPrehash(), attestationSigningKey.getPrivate());
    this.attestationVerificationKey = attestationSigningKey.getPublic();
    constructorCheck(attestationSigningKey.getPublic());
  }

  public SignedIdentityAttestation(byte[] derEncoding, AsymmetricKeyParameter verificationKey) throws IOException {
    ASN1InputStream input = new ASN1InputStream(derEncoding);
    ASN1Sequence asn1 = ASN1Sequence.getInstance(input.readObject());
    ASN1Sequence attestationEnc = ASN1Sequence.getInstance(asn1.getObjectAt(0));
    AlgorithmIdentifier algorithmEncoded = AlgorithmIdentifier.getInstance(asn1.getObjectAt(1));
    // TODO ideally this should be refactored to SignedAttestation being augmented with an generic
    // Attestation type and an encoder to construct such an attestation
    this.att = new IdentifierAttestation(attestationEnc.getEncoded());
    DERBitString signatureEnc = DERBitString.getInstance(asn1.getObjectAt(2));
    this.signature = signatureEnc.getBytes();
    this.attestationVerificationKey = verificationKey;
    if (!algorithmEncoded.equals(att.getSigningAlgorithm())) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Algorithm specified is not consistent"));
    }
    constructorCheck(verificationKey);
  }

  void constructorCheck(AsymmetricKeyParameter verificationKey) {
    if (!(verificationKey instanceof ECPublicKeyParameters)) {
      throw ExceptionUtil.throwException(logger,
          new UnsupportedOperationException("Attestations must be signed with ECDSA key"));
    }
    if (!verify()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Signature is not valid"));
    }
  }

  public IdentifierAttestation getUnsignedAttestation() {
    return att;
  }

  public byte[] getSignature() {
    return signature;
  }

  /**
   * Returns the public key of the attestation signer
   */
  public AsymmetricKeyParameter getAttestationVerificationKey() { return attestationVerificationKey; }

  @Override
  public byte[] getDerEncoding() {
    return constructSignedAttestation(this.att, this.signature);
  }

  static byte[] constructSignedAttestation(Attestation unsignedAtt, byte[] signature) {
    try {
      byte[] rawAtt = unsignedAtt.getPrehash();
      ASN1EncodableVector res = new ASN1EncodableVector();
      res.add(ASN1Primitive.fromByteArray(rawAtt));
      res.add(unsignedAtt.getSigningAlgorithm());
      res.add(new DERBitString(signature));
      return new DERSequence(res).getEncoded();
    } catch (Exception e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not encode asn1", e);
    }
  }

  @Override
  public boolean checkValidity() {
    return getUnsignedAttestation().checkValidity();
  }

  @Override
  public boolean verify() {
    try {
      if (!SignatureUtility.verifyEthereumSignature(att.getDerEncoding(), signature, attestationVerificationKey)) {
        logger.error("Could not verify signature");
        return false;
      }
    } catch (InvalidObjectException e) {
      logger.error("Could not decode the signature");
      return false;
    }
    return true;
  }

}

File: src/main/javascript/crypto/src/libs/SignedIdentityAttestation.ts
import {AsnParser} from "@peculiar/asn1-schema";
import {MyAttestation} from "../asn1/shemas/AttestationFramework";
import {KeyPair} from "./KeyPair";
import {hexStringToArray, uint8toBuffer, uint8tohex} from "./utils";
import {Attestation} from "./Attestation";
import {Verifiable} from "./Verifiable";
import {Validateable} from "./Validateable";
import {ASNEncodable} from "./ASNEncodable";
import {Asn1Der} from "./DerUtility";
import {IdentifierAttestation} from "./IdentifierAttestation";

export class SignedIdentityAttestation implements ASNEncodable, Verifiable, Validateable {
    private signature: string;
    private att: IdentifierAttestation;
    private commitment: Uint8Array;
    private uint8data: Uint8Array;
    private attestorKeys: KeyPair;
    static ECDSA_WITH_SHA256 = "1.2.840.10045.4.3.2";

    constructor() {}

    static fromBytes(uint8data: Uint8Array, attestorKeys: KeyPair): SignedIdentityAttestation {

        const myAttestation: MyAttestation = AsnParser.parse( uint8toBuffer( uint8data ), MyAttestation);
        return this.fromASNType(myAttestation, attestorKeys, uint8data);

    }

    static fromASNType(myAttestation: MyAttestation, attestorKeys: KeyPair, uint8data: Uint8Array = new Uint8Array(0)): SignedIdentityAttestation {
        let me = new this();
        me.uint8data = uint8data;
        me.attestorKeys = attestorKeys;
        let algorithmEncoded: string = myAttestation.signatureAlgorithm.algorithm;
        me.att = IdentifierAttestation.fromBytes(myAttestation.signedInfo) as IdentifierAttestation;

        // me.signature = myAttestation.signatureValue;
        me.signature = uint8tohex(new Uint8Array(myAttestation.signatureValue));
        if (algorithmEncoded !== me.att.getSigningAlgorithm()) {
            throw new Error("Algorithm specified is not consistent");
        }
        me.constructorCheck();
        return me;
    }

    static fromData(att: IdentifierAttestation, attestationSigningKey: KeyPair): SignedIdentityAttestation{
        let me = new this();
        me.attestorKeys = attestationSigningKey;
        me.att = att;
        // me.att.setSigningAlgorithm(SignedIdentityAttestation.ECDSA_WITH_SHA256);
        // me.signature = attestationSigningKey.signDeterministicSHA256( Array.from(me.att.getPrehash()));
        me.signature = me.attestorKeys.signRawBytesWithEthereum( Array.from(me.att.getPrehash()));
        me.constructorCheck();
        return me;
    }

    verify(){
        try {
            return this.attestorKeys.verifyBytesWithEthereum(hexStringToArray(this.att.getDerEncoding()), this.signature);

        } catch (e) {
            console.error(e);
            return false;
        }
    }

    checkValidity(){
        return this.getUnsignedAttestation().checkValidity();
    }

    getUnsignedAttestation(): IdentifierAttestation{
        return this.att;
    }

    getDerEncoding(): string{
        if (this.uint8data && this.uint8data.length){
            return uint8tohex(new Uint8Array(this.uint8data));
        } else {
            return this.constructSignedAttestation(this.getUnsignedAttestation(), this.signature);
        }
    }

    constructSignedAttestation(unsignedAtt: Attestation, signature: string){

        let rawAtt: Uint8Array = unsignedAtt.getPrehash();
        let alg = Asn1Der.encode('OBJECT_ID', unsignedAtt.getSigningAlgorithm());

        let res: string = uint8tohex(rawAtt) +
            Asn1Der.encode('SEQUENCE_30', alg) +
            Asn1Der.encode('BIT_STRING', '04' + uint8tohex(KeyPair.anySignatureToRawUint8(signature)));

        return Asn1Der.encode('SEQUENCE_30', res);
    }

    constructorCheck(){
        // TODO implement
        // if (!(verificationKey instanceof ECPublicKeyParameters)) {
        //     throw new UnsupportedOperationException("Attestations must be signed with ECDSA key");
        // }
        if (!this.verify()) {
            throw new Error("The signature is not valid");
        }
    }

}
